mailbox/postgres/src/test/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithmTest.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.ThreadIdGuessingAlgorithm;
+import org.apache.james.mailbox.store.mail.model.MimeMessageId;
+import org.apache.james.mailbox.store.mail.model.Subject;
+import org.apache.james.mailbox.store.quota.NoQuotaManager;
+import org.apache.james.metrics.tests.RecordingMetricFactory;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableSet;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.hash.Hashing;
+
+import reactor.core.publisher.Flux;
+
+public class PostgresThreadIdGuessingAlgorithmTest extends ThreadIdGuessingAlgorithmContract {
+    @RegisterExtension
+    static PostgresExtension postgresExtension = PostgresExtension.withRowLevelSecurity(PostgresMailboxAggregateModule.MODULE);
+
+    private PostgresMailboxManager mailboxManager;
+    private PostgresThreadDAO.Factory threadDAOFactory;
+
+    @Override
+    protected CombinationManagerTestSystem createTestingData() {
+        eventBus = new InVMEventBus(new InVmEventDelivery(new RecordingMetricFactory()), EventBusTestFixture.RETRY_BACKOFF_CONFIGURATION, new MemoryEventDeadLetters());
+        PostgresCombinationManagerTestSystem testSystem = (PostgresCombinationManagerTestSystem) PostgresCombinationManagerTestSystem.createTestingData(postgresExtension, new NoQuotaManager(), eventBus);
+        mailboxManager = (PostgresMailboxManager) testSystem.getMailboxManager();
+        messageIdFactory = new PostgresMessageId.Factory();
+        return testSystem;
+    }
+
+    @Override
+    protected ThreadIdGuessingAlgorithm initThreadIdGuessingAlgorithm(CombinationManagerTestSystem testingData) {
+        threadDAOFactory = new PostgresThreadDAO.Factory(postgresExtension.getExecutorFactory());
+        return new PostgresThreadIdGuessingAlgorithm(threadDAOFactory);
+    }
+
+    @Override
+    protected MessageMapper createMessageMapper(MailboxSession mailboxSession) {
+        return mailboxManager.getMapperFactory().createMessageMapper(mailboxSession);
+    }
+
+    @Override
+    protected MessageId initNewBasedMessageId() {
+        return messageIdFactory.generate();
+    }
+
+    @Override
+    protected MessageId initOtherBasedMessageId() {
+        return messageIdFactory.generate();
+    }
+
+    @Override
+    protected void saveThreadData(Username username, Set<MimeMessageId> mimeMessageIds, MessageId messageId, ThreadId threadId, Optional<Subject> baseSubject) {
+        PostgresThreadDAO threadDAO = threadDAOFactory.create(username.getDomainPart());
+        threadDAO.insertSome(username, hashMimeMessagesIds(mimeMessageIds), PostgresMessageId.class.cast(messageId), threadId, hashSubject(baseSubject)).block();
+    }
+
+    @Test
+    void givenAMailInAThreadThenGetThreadShouldReturnAListWithOnlyOneMessageIdInThatThread() throws MailboxException {
```

Comment:
Why those tests are not part of the contract?

Reply:
Instead of using threadDAO, getMessageIdsInThread method in CassandraThreadIdGuessingAlgorithm use mailboxManager to get data. The contract use messageMapper to create data for testing getMessageIdsInThread method. 
PostgresThreadIdGuessingAlgorithm use threadDAO to get data in getMessageIdsInThread method. Therefore, I also use threadDAO to create data for testing getMessageIdsInThread method. Thus, I have to override givenAMailInAThreadThenGetThreadShouldReturnAListWithOnlyOneMessageIdInThatThread of the contract